### ISSUE TITLE
Optional type hinting

### ISSUE THREAD

**Issue by jclc** [munificent/wren#519]
This is kind of an advanced feature that I don't expect to be incorporated any time soon, but I'd like to spark some discussion anyway. During my time developing on Python, I've found type hints to be a big convenience as they allow smart features in IDEs, such as type-aware autocompletion. I'm working on a static analyser/language server for Wren and this would allow it to provide some of these features. It would probably looks something like this:

```
class Foo {
  construct new() { // no type hinting required for constructor
    _bar = 0
  }
}

class FooStorage {
  static addFoo(foo) {
    if (__foos == null) {
      __foos = []
    }
    __foos.add(foo)
  }

  static hasFoo(foo): bool { // guaranteed to return a bool
    for (f in __foos) {
      if (f == foo) {
        return true
      }
    }
    return false
  }

  static getFooAt(index): Foo || null { // or ': optional[Foo]' but that would require a new keyword
    if (__foos == null || index < 0 || index >= __foos.count) {
      return null
    }
    return __foos[index]
  }

  static getFoos(): List[Foo] || List[Num] { // compound types
    if (__foos == null) {
      return [1, 1, 0, 3, 7]
    }
    return __foos
  }
}
```

The nice thing about doing this the same way as Python is that **the compiler doesn't have to care about any of this**; adhering to the type hints is entirely optional and can be ignored just like comments.

**Comment by jclc** [comment]
Here is a reference for how it's used in Python https://docs.python.org/3/library/typing.html

Python uses "->" for describing return values, but that's probably because Python blocks start with a colon, double use of which would have caused confusion there.

**Comment by heyitsmeuralex** [comment]
This is a cool idea. As a preliminary implementation we could use type hints as runtime assertions that check the types of arguments or returns as the program runs, and perhaps later magic static type-level stuff.

We should also consider `var foo: bool = true`.

**Comment by jclc** [comment]
> We should also consider var foo: bool = true.

I did have this in mind, particularly for function parameters, but left it out for the sake of brevity. That being said, If type hinting is used extensively in function declarations, I believe the need to specify the type after `var` becomes significantly less important as the type can be inferred from what is on the right side of the `=`. Of course there would be exceptions, for example when initialising a variable as null.

> Wrenalyzer was the start of this direction btw if you haven't seen.

I'm aware of this, but I don't think it quite suits my needs.

> How is your implementation going, and what approach have you been taking?

It's still very much work in progress, mainly because I haven't written a parser before. I'm taking cues from Wren's source code and Crafting Interpreters (thanks Bob!). I'm writing it in Go, since it makes binary distribution very easy and is practically designed for creating server applications.

My main motivation was the lack of good language support in Visual Studio Code. I'm aiming to at least have autocompletion and refactoring support with some basic problem reporting.

**Comment by jclc** [comment]
I would like to stress that what I want from type hinting is purely a tool for programming, not any kind of runtime or compile-time type assertions, like Bob seems to imply in that other issue. One of Wren's goals is to keep the actual VM and compiler simple and I don't intend to violate that.

You ought to be able to enforce correct types yourself if your static analyser warns you about incompatible types, so I don't see much point in implicit type assertions during runtime.

**Comment by underscorediscovery** [comment]
yea likewise, my intent was not a vm level distinction but a compile time one that happens before hand.

**Comment by mhermier** [comment]
https://github.com/munificent/wren/pull/527

**Comment by jclc** [comment]
We have preliminary support in the pull request, but we should still agree on some consistent rules for type annotations. When I suggested `||` to denote a union of multiple types, that was just the first thing that came to my mind. A two-byte token is a little unnecessary, though. Here's my revised suggestion:
```
// Forward slash for type unions, '?' for any type
import "foo" for Foo, Bar, Baz
class C {
    static f(): Foo/Bar/Baz/null {} // Foo, Bar, Baz or null.
    static g(): Map[Foo, Bar/Baz] {} // Map with Foos as keys and Bars and Bazs as values
    static h(): List[Bar/Baz]/List[Foo] {} // Either a list of Bars and Bazs or a list of Foos
    static i(): Map[String, ?] {} // Map with String keys and values of any type
}
```
I think `/` makes a lot of sense for unions, since it parallels its use in natural language (eg. "would you like some beer/wine/water?"). The pipe character `|` could also be used (just not my preference).

I don't think it's a good idea to add any new keyword solely for type annotations, it would potentially break existing code and I'm pretty sure @munificent wouldn't like that. For an 'any' type notation, `?` is the most intuitive option to me, but I would like to hear your opinions.

I'm basing a lot of this on Python's type annotations but I think we can keep this system much simpler. For example, I think 'optional' to signify a possible null return value is pretty unnecessary, you can just add `/null` or omit it entirely. Annotations would of course also apply to variable declarations and function parameters all the same.

**Comment by jclc** [comment]
List and Maps would be the only template-like types, though. If you don't know what types they're holding, annotating them isn't very useful in the first place. If I have a function that returns a list of instances of a certain class, but the annotation simply shows "List", I'd get type-aware completion for a list but nothing useful for any of its elements.

`Object` probably does make a lot more sense than `?`, that is a good point.

**Comment by mhermier** [comment]
I never said that hinting Map and List (or any other template like 
class) was not usefull, I only that that it would create a precedence in 
the language, regarding how to write templates. So it means we should 
put some extra care on choosing the syntax.

**Comment by jclc** [comment]
Sorry if I misunderstood you. I suppose I'm leaning towards `[,]` because of my Python bias, it is just such an elegant syntax. I suppose `<,>` would be more C++/Java-like, but that's not necessarily a virtue in my opinion; especially since `<` and `>` aren't used like brackets elsewhere in Wren. Of course, a tool could choose to support multiple notations as well. Since type annotations would be just white noise to the Wren compiler, it's not so much a language feature as it is a convention, right?

**Comment by heyitsmeuralex** [comment]
>  would be just white noise to the Wren compiler

If this is the case, the wren parser would need to gloss over all characters until it finds `{` or something denoting the end of that segment. I'd argue that's harder, and we're probably fine just having One True Way :tm: of writing wren type annotations.

**Comment by jclc** [comment]
That's true, we can't have any spurious curly brackets in there. I'd like to hear what Bob thinks about this, once he's got some time on his hands.

**Comment by mhermier** [comment]
There think there is some hope by extending '[]' to accept any number of
attributes. But it would add some ambiguity for user.

**Comment by jclc** [comment]
We would only need to support arbitrary amount of template arguments if we let users define their own template/compound classes. I'd be fine with just having that for Maps and Lists, for now at least.

**Comment by mhermier** [comment]
Well we can fake support of any number of argument like we did in the past,
by providing gazillion declaration manually written. It would be a good
start till we really need to support variadic argument.

**Comment by heyitsmeuralex** [comment]
So, as I understand it:

```
TypeHint = Type "/" TypeHint
         | Type

Type = Ident 
     | Ident "[" TemplateTypeList "]"

TemplateTypeList = TemplateTypeList "," Type
                 | Type
```

**Comment by mhermier** [comment]
Le lun. 30 avr. 2018 à 18:04, alex <notifications@github.com> a écrit :

> So, as I understand it:
>
> TypeHint = Type "/" TypeHint
>          | Type
>
> We can use '|' instead, it is allowed to override it, it is not reserved
by the compiler.
